zend_abstract_interface/zai_string/string.h

```diff
@@ -167,4 +172,82 @@ bool zai_option_str_get(zai_option_str self, zai_str *view) {
     return self.ptr;
 }
 
+/**
+ * Represents a owned string.
+ *
+ * When initializing the struct, use one of the initialization macros or
+ * functions. Do not initialize the struct directly.
+ *
+ * It is currently guaranteed that zai_string has the same layout as zai_str.
+ * This makes it safe to memcpy a zai_string into a zai_str. However, you
+ * shouldn't re-interpret a zai_string as a zai_str, because you are likely to
+ * commit an aliasing violation.
+ */
+typedef struct zai_string_s {
+    /**
+     * Guaranteed to hold .len + 1 bytes. The byte at self.ptr[self.len] is
+     * guaranteed to be a null character, but after initialization, this last
+     * byte must only be read, never written.
+     */
+    char *ptr;
+    size_t len;
+} zai_string;
+
+#define ZAI_STRING_EMPTY \
+    (zai_string) {.ptr = ZAI_STRING_EMPTY_PTR, .len = 0}
+
+/**
+ * Creates a zai_str view of the zai_string. Make sure the zai_str does not
+ * outlive the zai_string.
+ */
+__attribute__((pure))
+static inline zai_str zai_string_as_str(const zai_string *string) {
+    zai_str str;
+    memcpy(&str, string, sizeof(zai_str));
+    return str;
+}
+
+static inline zai_string zai_string_ctor(zai_str str) {
```

Comment:
I propose `zai_string_from_str()` (in case one wants a `zai_string_from_raw(const char *, size_t)` in future).
Similarly `zai_string_free()` instead of `zai_string_dtor()` for a better C terminology.

Reply:
I agree that `zai_string_from_str` makes sense over `zai_string_ctor`, but I think `dtor` should be kept because it's not freeing the pointer, it's freeing the interior contents of the string. You could also have `zai_string_free` and it would make sense:

```c
inline void zai_string_free(zai_string *string) {
    if (string != NULL) {
        zai_string_dtor(string);
        free(string);
    }
}
```

But we don't use that anywhere today, so it doesn't exist.

Reply:
You are probably right - it's destruction vs freeing. (Maybe `zai_string_destroy`?)

Reply:
I renamed it to `zai_string_destroy`.